### PR TITLE
fix(llmobs): sync and async generators can be decorated

### PR DESCRIPTION
**Pull request by sabrenner**
MLOB-1892

## What does this PR do
Fixes support for decorating sync and async generator functions. 

Some notes:

- While we will still capture function name and inputs (inputs only for non `llm` and `embedding` decorators), we will not automatically capture outputs from the generator
  - This change is intended to allow user annotation to still be captured, as previously, the generator would be captured as the "output", and user annotation would never go through.
- Async generators are not identifiable by `iscoroutine` and `isgeneratorfunction`, so we rely on `isasyncgenfunction` instead, verified in tests.

## Checklist
- [x] PR author has checked that all the criteria below are met
- The PR description includes an overview of the change
- The PR description articulates the motivation for the change
- The change includes tests OR the PR description describes a testing strategy
- The PR description notes risks associated with the change, if any
- Newly-added code is easy to change
- The change follows the [library release note guidelines](https://ddtrace.readthedocs.io/en/stable/releasenotes.html)
- The change includes or references documentation updates if necessary
- Backport labels are set (if [applicable](https://ddtrace.readthedocs.io/en/latest/contributing.html#backporting))

## Reviewer Checklist
- [x] Reviewer has checked that all the criteria below are met 
- Title is accurate
- All changes are related to the pull request's stated goal
- Avoids breaking [API](https://ddtrace.readthedocs.io/en/stable/versioning.html#interfaces) changes
- Testing strategy adequately addresses listed risks
- Newly-added code is easy to change
- Release note makes sense to a user of the library
- If necessary, author has acknowledged and discussed the performance implications of this PR as reported in the benchmarks PR comment
- Backport labels are set in a manner that is consistent with the [release branch maintenance policy](https://ddtrace.readthedocs.io/en/latest/contributing.html#backporting)

### CODE REVIEW

File: ddtrace/llmobs/decorators.py [hunk 1]
```diff
@@ -48,17 +79,32 @@ async def wrapper(*args, **kwargs):
 
             else:
 
+                @wraps(func)
+                def generator_wrapper(*args, **kwargs):
+                    if not LLMObs.enabled:
+                        log.warning(SPAN_START_WHILE_DISABLED_WARNING)
+                        yield from func(*args, **kwargs)
```
Comment (lievan) [inline ddtrace/llmobs/decorators.py:86]: nice, so `yield from` already handles `throw`, `send`, `GeneratorExit`, `StopIteration`?

File: tests/llmobs/test_llmobs_decorators.py [hunk 2]
```diff
@@ -536,3 +536,121 @@ def f(prompt, arg_2, kwarg_1=None, kwarg_2=None):
                 output_value="my custom output",
             )
         )
+
```
Comment (lievan) [inline tests/llmobs/test_llmobs_decorators.py:539]: should we add some test cases to ensure decorator generators are still compatible with `send`, `throw`..?

Comment (sabrenner) [reply tests/llmobs/test_llmobs_decorators.py:539]: good callout, let me try it out manually and then add a test if needed!

File: ddtrace/llmobs/decorators.py (hunk 1 above)
Comment (sabrenner) [reply ddtrace/llmobs/decorators.py:86]: yeah, it does (from some testing i've done) for sync generators (`yield from` can only be used in sync cases). the `async` versions are a bit different, will push some tests & fixes up today trying to handle `async` generators better for those cases.

File: ddtrace/llmobs/decorators.py [hunk 3]
```diff
@@ -13,6 +16,44 @@
...
+            except GeneratorExit:
+                await gen.aclose()
+                break
+            except StopAsyncIteration as e:
+                await gen.athrow(e)
+                break
+            except Exception as e:
+                await gen.athrow(e)
+                raise
+    except (StopAsyncIteration, GeneratorExit):
```
[the remark this answers is not in the document]
Comment (lievan) [reply ddtrace/llmobs/decorators.py:48]: By this line?
`next_val = await gen.asend(None)`

**Review by lievan — APPROVED** [review]
Thanks!

File: ddtrace/llmobs/decorators.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (sabrenner) [reply ddtrace/llmobs/decorators.py:48]: yeah, from that line. i special-cased these so we don't tag them as errors (as they are just generator lifecycle end indicators)

**Pull request merged**